src/combinatorics/simple_graph.lean

```diff
@@ -0,0 +1,176 @@
+/-
+Copyright (c) 2020 Aaron Anderson, Jalex Stark, Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Aaron Anderson, Jalex Stark, Kyle Miller.
+-/
+import data.fintype.basic
+import data.sym2
+
+open finset
+
+/-!
+# Simple graphs
+
+This module defines simple graphs on a vertex type `V` as an
+irreflexive symmetric relation.
+
+There is a basic API for locally finite graphs and for graphs with
+finitely many vertices.
+
+- A locally finite graph is one with instances `∀ v, fintype (G.adj v)`.
+
+- Given instances `decidable_rel G.adj` and `fintype V`, then the graph
+is locally finite, too.
+
+## Implementation notes
+
+TODO: This is the simplest notion of an unoriented graph.  This should
+eventually fit into a more complete combinatorics hierarchy which
+includes multigraphs and directed graphs.  We begin with simple graphs
+in order to start learning what the combinatorics hierarchy should
+look like.
+
+TODO: Part of this would include defining, for example, subgraphs of a
+simple graph.
+
+-/
+
+universe u
+
+/--
+A simple graph is an irreflexive symmetric relation `adj` on a vertex type `V`.
+The relation describes which pairs of vertices are adjacent.
+There is exactly one edge for every pair of adjacent edges;
+see `simple_graph.E` for the corresponding type of edges.
+
+Note: The type of the relation is given as `V → set V` rather than
+`V → V → Prop` so that, given vertices `v` and `w`, then `w ∈ G.adj v`
+works as another way to write `G.adj v w`.  Otherwise Lean cannot find
+a `has_mem` instance.
+-/
+structure simple_graph (V : Type u) :=
+(adj : V → set V)
+(sym : symmetric adj . obviously)
+(loopless : irreflexive adj . obviously)
+
+/--
+The complete graph on a type `V` is the simple graph with all pairs of distinct vertices adjacent.
+-/
+def complete_graph (V : Type u) : simple_graph V :=
+{ adj := λ v w, v ≠ w }
+
+instance (V : Type u) : inhabited (simple_graph V) :=
+⟨complete_graph V⟩
+
+instance complete_graph_adj_decidable (V : Type u) [decidable_eq V] :
+  decidable_rel (complete_graph V).adj :=
+by { dsimp [complete_graph], apply_instance }
+
+namespace simple_graph
+variables {V : Type u} (G : simple_graph V)
+
+/--
+The edges of G consist of the unordered pairs of vertices related by
+`G.adj`.  It is given as a subtype of the symmetric square.
+-/
+def E : Type u := {x : sym2 V // x ∈ sym2.from_rel G.sym}
```

Comment:
Again, I'm not super keen on this, because it requires everyone to "know the definition" to work with it (i.e. to know that you're meant to use it as a subtype).

Could we add a `has_mem` instance, provide a few iff lemmas, then mark is as irreducible before continuing further?

Reply:
(i.e. so the "implementation detail" involving `subtype` and `sym2` is completely hidden from the rest of the file, and hence from any users of the graph API?)

Reply:
(I'll admit to being picky here --- I have a new found enthusiasm for `@[irreducible]`, but it's okay to shout me down.)

Reply:
We did something here. It's possible that we still need another lemma or two before the `irreducible` attribute is declared, but we won't know what they are until we use the current API more.

Reply:
I was away this week on vacation, so I'm now getting to review the changes.  Everything's seemed good, but this one I'm not sure about.  My preference is for `E` to have the definition
```
def E : set (sym2 V) := sym2.from_rel G.sym
```
and I'm not sure exactly at which point it became a subtype. This is essentially the classical definition -- the edge set is a subset of unordered pairs on the vertex set.

If we make `E` irreducible, then we'd eventually be copying the whole interface from `data.sym2` (most of the interface for edges, except for `adj_iff_exists_edge`, are thin wrappers), and we'd have been better off defining edges from scratch in this file.  A reason not to do that is that, for example, multigraphs are likely going to use `sym2` as well in some capacity.